Subcell add function to compute TCI status and compute tag

nilsdeppe opened the pull request:
## Proposed changes

As per name...

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add function to compute the TCI status as a DataVector

Observing the TCI status is really useful for debugging limiters. The added
function computes the TCI status, setting it to 1 when the subcell scheme is
used, and to 0 otherwise.

nilsdeppe committed:
Add TciStatus compute tag

fmahebert commented on src/Evolution/DgSubcell/TciStatus.hpp:
```diff
@@ -0,0 +1,38 @@
+namespace evolution::dg::subcell {
+// @{
+/// Set `status` to `1` if the cell is marked as needing subcell, otherwise set
+/// `status` to `0`. `status` has grid points on the currently active mesh.
```
I have some confusion about the status... probably coming from some confusion about the relationship between `active_grid` and `tci_history.front()`? Some specific questions (reply in Github comment for now):
- if the status's number of grid points are set by `active_grid`, why can the status's value set by the history?
- if the `active_grid` is the desired source of the 0/1 value when the history is empty, why is it not when the history has contents?
- in what context might the value from the history disagree with the mesh. In other words, when would the `active_grid` be subcell but the history says we're on DG, or vice versa?

nilsdeppe replied:
You ask good questions, sir!

The TCI status is recorded in the history for LMM steppers, while for RK (and ADER in the future) there is no history for the TCI because I'm only allowing switching from subcell back to DG at full steps. The status lives on the _active grid_, but the status is determined by the history if we are doing LMM stepping. This is because we are waiting for the shock to clear the stepper history. Does that answer all the questions?

fmahebert replied:
New question: if the status is determined by the history (as you said: waiting for the shock to clear the history), why is the status set by a single element `history.front()` rather than a check over all the history data that any of them are subcell?

nilsdeppe replied:
Because it is the most recent element that determines what the TCI decided at this time level. It is the status of the TCI, and not the status of the history of the TCI. I consider the history portion to be an artifact of the time stepper used, not part of the TCI that determines if the immediate solution could safely be projected to DG.

fmahebert replied:
New question: if the status is an indication of the state of the solution "right now", why does it sometime read from `active_grid` and sometimes from `history`? I understand there's no history if you're using a substep/ader method, but why not use `active_grid` always?

nilsdeppe replied:
because if the history says that there is still a shock, the active grid will still be subcell, even though the TCI has deemed the current time level safe. That is, the current time level is fine, but we would do bad things because the history isn't fine. Thus, the active grid is not indicative of what the TCI has decided, but rather indicative of what the TCI+time stepper history requires

kidder replied:
my confusion is, if the active grid is subcell, but the first item in the history is dg, why is the status set to 0 ?

fmahebert replied:
Oh I see. So you don't generally expect cases where the history says "we're shock-y now" and the active_grid is DG (this sounds like a bug), whereas you DO expect cases where the history says "we're smooth now" and active_grid is Subcell (because of shock-y earlier history). Correct?

(edit: typed this before seeing Larry's question, so treat as independent)

nilsdeppe replied:
Yes, that's right. Here's a simple 2nd order LMM case:

We are running on subcell...
- t^n: TCI decides bad, record TCI status as `Subcell` grid, and status is `1` at `t^n`
- t^{n+1}: TCI decides all is good, record TCI status as `Dg` grid, and TCI status is `0` at `t^{n+1}`. However, since we are using a 2nd Adams-Bashforth method we need to wait for the previous entry in the history to also be `Dg`. Since `t^n` is `Subcell` the `active_grid` must remain `Subcell`
- t^{n+2}: TCI decides all is good, record TCI status as `Dg` grid, and TCI status is `0` at `t^{n+2}`. _Now_ we can switch back to DG since the entire time stepper history is `Dg`.

When using DG we won't even be tracking the history because we can always switch back to subcell, it's just switching back to DG that needs to be restricted. Does this make sense now?

Re @kidder: because the TCI decided the solution is valid, but we can't do time stepping with an LMM because the time stepper history contains a discontinuity. The TCI history really is what the TCI decided, but the active grid is not chosen _solely_ on the TCI decision at the current time, but also the TCI decision in the past.

kidder replied:
Okay, I'm typing this out in detail  because I am obviously confused about something...

-  We record the TCI status on the current active grid
-  The TCI status is determined completely from current information
-  The current active grid is determined based on the current TCI status and (for an `N`th-order multi-step method) the  `N-1` most recent TCI statuses 
- `tci_status` is called with the current `active_grid` as determined by the description of the previous bullet and the history of the previous `N-1` (and now I see part of my confusion, I don't know what tci_history holds... is it the active_grid of the past steps or the tci_status recorded using Dg for a 0 and Subcell for a 1?). And for clarification is tci_history.front() the oldest or latest time in the history?  I was assuming oldest...

I can't go further until those questions are answered and you tell me if any of my bullets are incorrect...

fmahebert replied:
I suggest this addition to the documentation (feel free to edit):
```
/// \note It is possible to encounter a `status` of `0`, indicating a smooth
/// solution, even though the active mesh is the subcell mesh. This can occur
/// when using a multistep time integration method like Adams Bashforth, where
/// the FD/FV scheme is used as long as _any_ of the timestepper history is
/// flagged for FD/FV subcell evolution. An example of this woul be just after a
/// shock leaves the cell: the solution is now smooth (`status` takes value
/// `0`), but the multistep method continues to take FD timesteps (`status` is
/// defined on the subcell mesh) until the entire history is flagged as smooth.
```

Possible additional suggestion is to make the opposite case DG mesh but TCI triggers an error of some kind... what do you think, is this reasonable?

(edit: sorry, again, typing in parallel to Larry. feel free to ignore my suggestion until we're all on board)

fmahebert commented on tests/Unit/Evolution/DgSubcell/Test_Tags.cpp:
```diff
@@ -58,6 +58,9 @@ void test() {
             logical_coords_box) == logical_coordinates(subcell_mesh));
+
+  TestHelpers::db::test_compute_tag<
+      evolution::dg::subcell::Tags::TciStatusCompute<Dim>>("TciStatus");
```
Is the test of the corresponding simple tag missing or am I just not seeing it?

nilsdeppe replied:
Huh, apparently that got lost? I'll add it :)

nilsdeppe replied:
I think I goofed when I replied here... IT appears that the simple tag is tested on line 80 (new) below. The simple tag doesn't depend on dimension :)

kidder commented on src/Evolution/DgSubcell/TciStatus.cpp:
```diff
@@ -0,0 +1,70 @@
+
+namespace evolution::dg::subcell {
+template <size_t Dim>
+void tci_status(const gsl::not_null<Scalar<DataVector>*> status,
```
include DataVector

kidder reviewed (commented):
I have no additional questions or requested changes

nilsdeppe commented:
I've rebased and pushed a fixup expanding slightly on what @fmahebert suggested. Please let me know if further changes would help clarify things more :)

kidder commented on src/Evolution/DgSubcell/TciStatus.hpp:
```diff
@@ -23,6 +23,18 @@ namespace evolution::dg::subcell {
+/// a shock leaves the cell: the solution is now smooth (`status` takes value
+/// `0`), but the multistep method continues to take FD timesteps (`status` is
+/// defined on the subcell mesh) until the entire history is flagged as smooth.
+/// Thus, he `tci_history.front()` is used to set the status when using
```
I would suggest `Thus tci_history.front() (which corresponds to the most recent tci status) is used...`

kidder reviewed (commented):
I am fine for you to squash

nilsdeppe commented:
Rebased and pushed another fixup, thanks! :)

kidder reviewed (commented):
you have my permission to squash

fmahebert reviewed (commented):
Either the `test_simple_tag` is still missing, or you've added it and I still can't find it...

I am happy with the documentation clarifications in the fixups, so you can squash them and the above change.

nilsdeppe commented:
I've rebased and squashed. I also realized that the simple tag for TciStatus is tested outside the dimension-dependent function, sorry :(

Thanks for the reviews :)
